apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/json/MapJsonReader.kt

```diff
@@ -23,285 +24,298 @@ import com.apollographql.apollo3.api.json.internal.toLongExact
-    return when (val data = dataStack[stackSize - 1]) {
-      is List<*> -> {
-        data[indexStack[stackSize - 1]]
+      peekedToken = when (next) {
+        is Map.Entry<*, *> -> JsonReader.Token.NAME
+        else -> anyToToken(next)
       }
-      is OrderedMap -> {
-        check(pendingName != null)
-        data.entries[indexStack[stackSize - 1]].value
+    } else {
+      peekedToken = if (path[stackSize - 1] is Int) {
+        JsonReader.Token.END_ARRAY
+      } else {
+        JsonReader.Token.END_OBJECT
       }
-      else -> error("")
     }
   }
 
-  private fun consumeValue() {
-    pendingName = null
-    indexStack[stackSize - 1]++
-    incrementPathIndex()
-  }
-
   override fun beginArray() = apply {
-    path += "."
-    path += -1
-    val list = nextValue()
-    check(list is List<*>)
-    consumeValue()
-    push(list)
+    if (peek() != JsonReader.Token.BEGIN_ARRAY) {
+      throw JsonDataException("Expected BEGIN_ARRAY but was ${peek()} at path ${getPath()}")
+    }
+
+    val currentValue = peekedData as List<Any?>
+
+    check(stackSize < MAX_STACK_SIZE) {
+      "Nesting too deep"
+    }
+    stackSize++
+
+    path[stackSize - 1] = 0
+    iteratorStack[stackSize - 1] = currentValue.iterator()
+    advanceIterator()
   }
 
   override fun endArray() = apply {
-    pop()
-    rewindPath()
-    incrementPathIndex()
+    if (peek() != JsonReader.Token.END_ARRAY) {
+      throw JsonDataException("Expected END_ARRAY but was ${peek()} at path ${getPath()}")
+    }
+    stackSize--
+    iteratorStack[stackSize] = null // allow garbage collection
+    path[stackSize] = null // allow garbage collection
+    advanceIterator()
   }
 
   override fun beginObject() = apply {
-    path += "."
-    val map = nextValue()
-    check(map is Map<*, *>)
-    consumeValue()
-    push(OrderedMap(map.entries.map { Entry(it.key as String, it.value) }))
-  }
+    if (peek() != JsonReader.Token.BEGIN_OBJECT) {
+      throw JsonDataException("Expected BEGIN_OBJECT but was ${peek()} at path ${getPath()}")
+    }
 
-  override fun endObject() = apply {
-    pop()
-    rewindPath()
-    incrementPathIndex()
+    @Suppress("UNCHECKED_CAST")
+    container = peekedData as Map<String, Any?>
+
+    check(stackSize < MAX_STACK_SIZE) {
+      "Nesting too deep"
+    }
+    stackSize++
+
+    rewind()
   }
 
-  private fun anyToToken(any: Any?) = when (any) {
-    null -> JsonReader.Token.NULL
-    is List<*> -> JsonReader.Token.BEGIN_ARRAY
-    is Map<*, *> -> JsonReader.Token.BEGIN_OBJECT
-    is Int -> JsonReader.Token.NUMBER
-    is Long -> JsonReader.Token.NUMBER
-    is Double -> JsonReader.Token.NUMBER
-    is String -> JsonReader.Token.STRING
-    is Boolean -> JsonReader.Token.BOOLEAN
-    else -> error("Unsupported value $any")
+  override fun endObject() = apply {
+    if (peek() != JsonReader.Token.END_OBJECT) {
+      throw JsonDataException("Expected END_OBJECT but was ${peek()} at path ${getPath()}")
```

Comment:
This is more strict than the previous implementation and will fail on consumers that do not consume all tokens until the end of the object (like BufferedSourceJsonReader)